<h1>Анализ базы данных книжного сервиса</h1>

**Описание проекта:**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. На какое-то время жители городов перестали выходить на улицу, посещать кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задачи проекта:**

1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитать количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц (исключить из анализа брошюры);
4. Определить автора с самой высокой средней оценкой книг (только книги с 50 и более оценками);
5. Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Описание данных:**

Таблица **`books`** содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица **`authors`** cодержит данные об авторах

* `author_id` — идентификатор автора;
* `author` — имя автора.

Таблица **`publishers`** cодержит данные об издательствах:

* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

Таблица **`ratings`** cодержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица **`reviews`** cодержит данные о пользовательских обзорах:

* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

### Подключение к базе данных

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
from IPython.display import Image, display

In [ ]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args = {'sslmode':'require'})

**Схема данных:**

In [ ]:
display(Image(url = 'https://pictures.s3.yandex.net/resources/scheme_1589269096.png'))    

### Исследование данных по таблицам

In [ ]:
query = '''SELECT * FROM books LIMIT 10'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [ ]:
query = '''SELECT * FROM authors LIMIT 10'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [ ]:
query = '''SELECT * FROM publishers LIMIT 10'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [ ]:
query = '''SELECT * FROM ratings LIMIT 10'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [ ]:
query = '''SELECT * FROM reviews LIMIT 10'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

### Подсчет количества книг вышедших после 1 января 2000 года

In [ ]:
query = '''SELECT COUNT(book_id)
           FROM books
           WHERE publication_date > CAST('2000-01-01' as date)'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

count
0    819

* Количество книг вышедших после 1 января 2000 года - **819**

### Расчет количества обзоров и средней оценки для каждой книги

In [ ]:
query = '''WITH br AS (
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rw.review_id) AS review_count,
       AVG(rt.rating) AS average_rating
FROM books b
LEFT JOIN reviews rw ON b.book_id = rw.book_id
LEFT JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id, b.title
)
SELECT br.title, 
       br.review_count, 
       br.average_rating
FROM br
ORDER BY br.review_count'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

title  review_count  \
0      Anne Rice's The Vampire Lestat: A Graphic Novel             0   
1     The Cat in the Hat and Other Dr. Seuss Favorites             0   
2                            Essential Tales and Poems             0   
3                              The Natural Way to Draw             0   
4    Disney's Beauty and the Beast (A Little Golden...             0   
..                                                 ...           ...   
995                          The Giver (The Giver  #1)             6   
996                                           The Road             6   
997  The Curious Incident of the Dog in the Night-Time             6   
998             The Da Vinci Code (Robert Langdon  #2)             6   
999                            Twilight (Twilight  #1)             7   

     average_rating  
0          3.666667  
1          5.000000  
2          4.000000  
3          3.000000  
4          4.000000  
..              ...  
995        3.750000  
996        3.772727  
997        4.081081  
998        3.830508  
999        3.662500  

[1000 rows x 3 columns]

* Для каждой книги посчитано количество обзоров и средний рейтинг. Больше всего обзоров на книгу - Twilight (Twilight #1).	

### Опредение издательства, с наибольшим числом книг толще 50 страниц

In [ ]:
query = '''
SELECT p.publisher,
       COUNT(b.book_id) AS book_count
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY book_count DESC
LIMIT 1'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

publisher  book_count
0  Penguin Books          42

* Издательство **Penguin Books** выпустило наибольшее число книг толщиной больше 50 страниц - **42** книги.

### Определение автора, с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [ ]:
query = '''WITH ar AS ( 
SELECT book_id,
       COUNT(rating_id) 
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) >= 50
)
SELECT author,
       AVG(rating) AS average_rating
FROM books b
JOIN ratings r ON b.book_id = r.book_id 
JOIN authors a ON b.author_id = a.author_id
WHERE b.book_id IN (SELECT book_id FROM ar)
GROUP BY a.author
ORDER BY average_rating DESC
LIMIT 1'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)


author  average_rating
0  J.K. Rowling/Mary GrandPré        4.287097

* Автор c самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré**, средняя оценка книг автора - **4.28** балла.

### Подсчет среднего количества обзоров от пользователей, которые поставили больше 48 оценок

In [ ]:
query = '''WITH users_48 AS (
SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) > 48
)
SELECT AVG(rw_count) AS avg_rw_per_user
FROM (SELECT COUNT(review_id) AS rw_count
      FROM reviews
      WHERE username IN (SELECT username FROM users_48)
      GROUP BY username
) AS user_rw'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(query), con = con)

avg_rw_per_user
0             24.0

* Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - **24** обзора.

### Результаты анализа базы данных книжного сервиса:

1. Количество книг вышедших после 1 января 2000 года - **819**.
2. Для каждой книги посчитано количество обзоров и средний рейтинг. Больше всего обзоров на книгу - Twilight (Twilight #1).
3. Издательство **Penguin Books** выпустило наибольшее число книг толщиной больше 50 страниц - **42 книги**.
4. Автор c самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré**, средняя оценка книг автора - **4.28 балла**.
5. Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - **24 обзора**.